crates/precompile/src/bn128/substrate.rs

```diff
@@ -129,6 +129,33 @@ pub(super) fn read_g2_point(input: &[u8]) -> Result<G2, PrecompileError> {
     new_g2_point(ba, bb)
 }
 
+/// Reads a scalar from the input slice
+///
+/// Note: The scalar does not need to be canonical.
+#[inline]
+pub(super) fn read_scalar(input: &[u8]) -> bn::Fr {
+    assert_eq!(
+        input.len(),
+        SCALAR_LEN,
+        "unexpected scalar length. got {}, expected {SCALAR_LEN}",
+        input.len()
+    );
+    // `Fr::from_slice` can only fail when the length is not `SCALAR_LEN`.
+    bn::Fr::from_slice(input).unwrap()
```

Comment:
Originally this was returning an error, however we would need to add a new enum variant into PrecompileError for ScalarLength invalid. This would be a breaking change for those who are matching on enums so I chose to do what was already happening (panicking when length is not 32)